src/Sarif.Multitool/Rules/SARIF1004.ExpressUriBaseIdsCorrectly.cs

```diff
@@ -63,14 +70,48 @@ private void AnalyzeOriginalUriBaseIdsEntry(string uriBaseId, ArtifactLocation a
             {
                 var uri = new Uri(uriString, UriKind.RelativeOrAbsolute);
 
+                // TopLevelUriBaseIdMustBeAbsolute: Top level uriBaseId must be absolute.
                 if (artifactLocation.UriBaseId == null && !uri.IsAbsoluteUri)
                 {
-                    LogResult(pointer, nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_TopLevelUriBaseIdMustBeAbsolute_Text), uriString, uriBaseId);
+                    // {0}: The URI '{1}' belonging to the '{2}' element of run.originalUriBaseIds is not an absolute URI.
+                    LogResult(
+                        pointer,
+                        nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_TopLevelUriBaseIdMustBeAbsolute_Text),
+                        uriString,
+                        uriBaseId);
                 }
 
+                // UriBaseIdValueMustEndWithSlash: uriBaseIds must end with a slash.
                 if (!uriString.EndsWith("/"))
                 {
-                    LogResult(pointer, nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustEndWithSlash_Text), uriString, uriBaseId);
+                    // {0}: The URI '{1}' belonging to the '{2}' element of run.originalUriBaseIds does not end with a slash.
+                    LogResult(
+                        pointer,
+                        nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustEndWithSlash_Text),
+                        uriString,
+                        uriBaseId);
+                }
+
+                // UriBaseIdValueMustNotContainDotDotSegment: uriBaseIds must not contain `..` segment(s).
+                if (uriString.Contains(".."))
+                {
+                    // {0}: '{1}' '{2}' Placeholder: SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustNotContainDotDotSegment_Text
+                    LogResult(
+                        pointer,
+                        nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustNotContainDotDotSegment_Text),
+                        uriString,
+                        uriBaseId);
+                }
+
+                // UriBaseIdValueMustNotContainQueryOrFragment: uriBaseIds must not contain any query or fragments.
+                if (uri.IsAbsoluteUri && (!string.IsNullOrEmpty(uri.Fragment) || !string.IsNullOrEmpty(uri.Query)))
```

Comment:
>uri.IsAbsoluteUri [](start = 20, length = 17)

Yes, I'm afraid that's the best you can do. If there are relative URIs with path or query you won't see them but I don't think it's worth trying to parse on your own.

Reply:
i can do the same thing with segments?

---
In reply to: [445146586](https://github.com/microsoft/sarif-sdk/pull/1930#discussion_r445146586) [](ancestors = 445146586)

Reply:
I don't recommend it. Unlike query and fragment parsing, it's really easy to detect .. segments (split on /). I'm willing to do that by hand.

---
In reply to: [445148607](https://github.com/microsoft/sarif-sdk/pull/1930#discussion_r445148607) [](ancestors = 445148607,445146586)

Reply:
yes, and there is higher chance someone has a relative uri with \..\.. vs queries/fragements.. so we should let them know sooner.

---
In reply to: [445149445](https://github.com/microsoft/sarif-sdk/pull/1930#discussion_r445149445) [](ancestors = 445149445,445148607,445146586)